### PR TITLE
79 transition to real game

### PR DESCRIPTION
**Pull request by FedericoDM**
## Describe Your Changes

- 'Round' endpoint also leads to 'end of game' template
- Added 'Practice Game' or 'Real Game' to the 'round' template
- Created `real_game_transition.html` template

Transition screen successful and looks like this:

image

Practice or Real game now displayed at the top right of the screen:

image

If the following game is not practice, students are redirected to the default game start page:

image

### CODE REVIEW

**Comment by FedericoDM** [comment]
@anisfeld I fixed an issue where the 'round' endpoint didn't lead to the end of game. This was necessary for the 'session' loop that we will be having,

I am still a bit lost, however, on how to lead them from the practice game (or any game) to the following. 

Any thoughts?

**Comment by anisfeld** [comment]
Good catch on stopping the loop! 

I'm not sure what particular part of this you are stuck on. 


# Testing for end of practice

All the integrated/segregated games are added to the database at session creation and the practice game(s) are noted as practice with a boolean in the database.

So the game table is heuristically looks like: 

| id | next_game_id| Practice | etc |
|--------|--------|--------|--------|
| 1 | 2| True | ... |
| 2 | 3 | True | ...|
| 3 | 4 | False | ... | 

So then, we need a test at the end of each game `is_end_of_practice()` that  checks if the current game is practice and if so, if the next game is *not* practice.  If so, we redirect the user to the welcome to the real game routes. 

# Game flow once test passes/fails

I think the place to put the test is in the `go_to_next_game` route. If is_end_of_practice() tell them that practice is over. 

I'm happy to discuss this further in real life.

**Comment by FedericoDM** [comment]
@anisfeld Ok that makes it a bit clearer! Right now, when the game ends, we are redirecting game to the 'end of game' screen. Will this still be the case? I assume that, after this screen, we will see if the current game is just a practice game and if the next game is not.


We can meet in person next week to discuss this.

**Comment by FedericoDM** [comment]
Transition screen looks like this:

image

**Comment by FedericoDM** [comment]
Transition into next game is successful, type of game is also displayed on the top right

image

**Comment by FedericoDM** [comment]
If the next game is not a practice game, it is redirected to the standard 'start_of_game' screen
image

**Comment by anisfeld** [comment]
I think your solution for ending the game isn't working ... This is a practice game at round 9:
Screenshot 2023-12-04 at 11 28 41 PM

**Comment by FedericoDM** [comment]
How many rounds does your game have? I had to reduce the number of rounds for practice and real games to 3 (it was set up to 10) @anisfeld

**Comment by anisfeld** [comment]
Ah. I didn't think it would be 10 rounds--my bad! 

The end of round screen should be different for a practice game as well. It should not say they're getting paid.

**Comment by FedericoDM** [comment]
Perfect @anisfeld! I'll make the changes for that and will let you know when they are done

**Comment by FedericoDM** [comment]
Done, @anisfeld.

The end_of_game screen now displays the following text when the finished game was a practice one.

image

File: prijateli_tree/app/routers/games.py [hunk 1]
```diff
@@ -339,6 +321,9 @@ def view_round(
     # Get current round
     if current_round == 1:
         template_data["ball"] = player.initial_ball
+    elif current_round > game.rounds:
+        redirect_url = f"/games/{game_id}/player/{player_id}/end_of_game"
```
Comment (anisfeld) [inline prijateli_tree/app/routers/games.py:325]: The redirect should be to the function that adds the score.  As a complication, we don't want to add a score for practice games.

**Review by anisfeld — CHANGES_REQUESTED** [review]
This looks good. Just need to add the score (during real rounds)!

**Comment by FedericoDM** [comment]
Added the redirect to the 'add points' @anisfeld

**Comment by anisfeld** [comment]
The redirect didn't work.  Test the code by playing the game when you make changes. I use a mixture of my database viewer (DBeaver), browser, server log and command line to make sure things work the way I think they do. We will try to transition to more test-driven development moving forward. 

It seems like redirecting from GET to POST is taboo. I used the javascript code in waiting  + the all_set API to do the score updating. This might also be a bad idea in the long run since the logic of the game is now straddling python and javascript. 

In the meantime, I made too many changes! Please go through the code and act as a reviewer.

**Comment by FedericoDM** [comment]
I have the same setup! I tested your changes and everything looks good, but I dont see any points being added to the db.

**Comment by FedericoDM** [comment]
Should be ready to go! @anisfeld

**Review by anisfeld — APPROVED** [review]

**Pull request merged**